[FEATURE] Ajouter et alimenter une colonne lastLoggedAt dans la table user-logins (PIX-9008)

lego-technix opened the pull request:
## :unicorn: Problème

Les informations relatives aux connexions/logins d'un utilisateur sont réparties sur 2 tables différentes ayant des fréquences de mise à jour différentes (`users` et `user-logins`, la dernière étant constamment modifiée par les différents logins alors que la première varie moins) au lieu d'être dans une table unique, ce qui : 
* complexifie la compréhension et le travail des développeurs
* augmente la contention de la base de données, ce qui diminue les performances de la base de données et augmente les risques de deadlock

## :robot: Proposition

Déplacer la colonne `lastLoggedAt` de la table `users` vers la table `user-logins`.

Ceci doit être fait en implémentant une solution rétro-compatible qui est la solution idéale en la matière.

Cette 1ère PR réalise les actions suivantes : 
1. Ajouter une migration pour créer la nouvelle colonne
2. Mettre à jour la nouvelle colonne partout où l’ancienne est mise à jour tout en conservant et continuant à modifier et à lire l’ancienne colonne

## :rainbow: Remarques

Cette PR fait partie d'un ensemble de PR.

## :100: Pour tester

* Se connecter avec succès un utilisateur ayant l'authentification Pix (email: `user-1-7003@example.net `, mot de passe par défaut). Constater que la date et l'heure de connexion correspondent bien à la tentative réussie :
   * dans PixAdmin (qui se base sur la table `users`)
   * dans la table `user-logins`
* Se connecter avec succès un utilisateur ayant une authentification SSO OIDC (par exemple avec Pôle Emploi, n'importe quel compte convient) et ne s'étant encore jamais connecté. Constater que la date et l'heure de connexion correspondent bien à la tentative réussie :
   * dans PixAdmin (qui se base sur la table `users`)
   * dans la table `user-logins`
* Se connecter avec succès un utilisateur ayant une authentification SSO OIDC (par exemple avec Pôle Emploi, n'importe quel compte convient) et s'étant déjà connecté. Constater que la date et l'heure de connexion correspondent bien à la tentative réussie :
   * dans PixAdmin (qui se base sur la table `users`)
   * dans la table `user-logins`
* Se connecter avec succès un utilisateur ayant une authentification SSO OIDC (par exemple avec Pôle Emploi, n'importe quel compte convient) dans le cas d'une réconciliation. Constater que la date et l'heure de connexion correspondent bien à la tentative réussie :
   * dans PixAdmin (qui se base sur la table `users`)
   * dans la table `user-logins`
* Se connecter avec succès un utilisateur ayant une authentification GAR (Saml ID: `SamlID`, Prénom: `Eliza`, Nom: `Delajungle`). Constater que la date et l'heure de connexion correspondent bien à la tentative réussie :
   * Entrer le code campagne `SCOBADGE1`
   * dans PixAdmin (qui se base sur la table `users`)
   * dans la table `user-logins`
* Se connecter avec succès un utilisateur ayant une authentification GAR dans le cas d'une réconcilation, c'est à dire un utilisateur ayant déjà une méthode d'authentification PIX (Saml ID: `SamlID`, Prénom: `Chihiro`, Nom: `Ogino`)
   * Entrer le code campagne `SCOBADGE1`
   * Remplir les informations de reconciliation (date de naissance: `01/01/2013`)
   * Constater l'affichage d'une popup indiquant l'existence d'un compte utilisateur correspondant aux meme identifiants
   * Cliquer sur le bouton `Continuer avec les identifiants PIX`
   * Constater l'affichage d'une double mire de connexion
   * Se connecter avec ses identifiants PIX (email: `chihiro.ogino@miyazaki.net`, mot de passe par défaut)
   * Constater l'affichage du tutoriel de campagne

lego-technix committed:
feat(api): create knex migration for adding new column lastLoggedAt in user-logins table

Co-authored-by: Clément Latzarus <clement.latzarus@pix.fr>
Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: BKsensei <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

er-lim committed:
feat(api): add updateLastLoggedAt method in user-login-repository

Co-authored-by: Clément Latzarus <clement.latzarus@pix.fr>
Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

clemlatz committed:
refactor(api): use constant for table name in test

igorissen committed:
feat(api): modify authenticate-user usecase

in order to update user login lastLoggedAt

Co-authored-by: Clément Latzarus <clement.latzarus@pix.fr>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

igorissen committed:
feat(api): modify authenticate-external-user usecase

in order to update user-login lastLoggedAt

Co-authored-by: Clément Latzarus <clement.latzarus@pix.fr>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr6952.review.pix.fr)
  * [App (.org)](https://app-pr6952.review.pix.org)
  * [Orga (.fr)](https://orga-pr6952.review.pix.fr)
  * [Orga (.org)](https://orga-pr6952.review.pix.org)
  * [Certif (.fr)](https://certif-pr6952.review.pix.fr)
  * [Certif (.org)](https://certif-pr6952.review.pix.org)
  * [1D](https://1d-pr6952.review.pix.fr)
  * [Admin](https://admin-pr6952.review.pix.fr)
  * [API](https://api-pr6952.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr6952.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr6952/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr6952/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr6952/environment)

igorissen committed:
feat(api): modify authenticate-oidc-user usecase

in order to update user login lastLoggedAt

Co-authored-by: Clément Latzarus <clement.latzarus@pix.fr>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

igorissen committed:
feat(api): modify reconcile-oidc-user usecase

in order to update user login lastLoggedAt

Co-authored-by: Clément Latzarus <clement.latzarus@pix.fr>
Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

Thomasevano committed:
feat(api): modify create-oidc-user usecase

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>

Thomasevano committed:
feat(api): modify create-user-and-reconcile-to-organization-learner-from-external-user usecase

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>

Thomasevano committed:
feat(api): modify get-external-authentication-redirection-url usecase

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>

lego-technix committed:
refactor(api): use constant for table name in user-login-repository

igorissen committed:
refactor(audit): use fake timers

Co-authored-by: Eric Lim <eric.lim@zenika.com>
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>

igorissen committed:
feat(api): add new organization learner

without GAR authentication method for usecase where GAR user already has a Pix account

igorissen committed:
refactor(api): code improvement

Thomasevano committed:
refactor(api): fix typo in test conventions

Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>
Co-authored-by: LEGO Technix <109212476+lego-technix@users.noreply.github.com
Co-authored-by: Rébecca Kaci <rebecca.kaci@gmail.com>
Co-authored-by: Eric Lim <eric.lim@zenika.com>

er-lim commented:
Tests fonctionnels OK sur Chrome ✅

lego-technix commented:
Tests fonctionnels OK sur Firefox ✅
